### ISSUE TITLE
Add the possibility for extensions to provide custom vars

### ISSUE THREAD

**Issue by mems** [adobe/reactor-turbine#116]
I would like to automatically provide some data element without the user to create data element for each possible items in each Launch properties.

My example is to be able to set Adobe Analytics settings as:

```
channel: "%digitalData.page.category.primaryCategory%"
propX: "%digitalData.page.attributes.marketID%"
propY: "%digitalData.user.0.segment.subscriptionOffer%"
prop2: "%window.location.href%"
```

For that, I use my extension main module:

```js
_satellite.setVar({
	window: window,
	digitalData: window.digitalData
});
```

## The main problem

A module of Adobe Analytics ([get-tracker](https://developer.adobelaunch.com/extensions/shared_modules/adobe-analytics/#get-tracker)) when get hydrated, synchronously get the AA extension settings (via `getExtensionSettings()`) that force to resolve all tokens in that settings, even before any other extensions had the opportunity to declare custom variables.

It isn't always the case works. Some times it's works because the extension list is in a different order.

> `main` (optional)
>
> [...]
> 
> This module will always be included in the runtime library and executed. Because the module is always included in the runtime library, we recommend only using a “main” module when absolutely necessary and keeping its code size minimal.
> 
> This module is not guaranteed to be executed first; other modules may be executed before it.
> 
> — [Adobe Experience Platform Launch](https://developer.adobelaunch.com/extensions/reference/extension-manifest/)

## Additional problem: non-browsable data elements

Data elements aren't browsable because it could break [backward compatibility](https://developer.adobelaunch.com/extensions/reference/backwards-compatibility/):

https://github.com/adobe/reactor-turbine/blob/2eec8bdb3ed6bd43f8bab53cfeaddfb5e5a6f34b/src/createGetVar.js#L82-L83

So you need to declare one data element per fields that exist in your datalayer, which generate a library that is quite heavy (with all `{modulePath:"path/to/module.js",settings:{...` per datalayer fields). And it's more difficult to maintains, especially when this configuration is used by multiple (~10) properties.

## Possible solutions

The following is what I think as possible solutions.

### Add a way for extensions to wait all modules are hydrated

With an API, it could be possible to execute something only after all modules has been hydrated. Like wait a promise `turbine.hydration` that resolve after hydration phase?

A simplier solution without a dedicated API, is to wait the next event cycle with `setTimeout(() => turbine.getExtensionSettings())`?
Something that the Adobe Analytics extension could use?

### Add a ways to declare custom vars via the manifest

But I'm not sure if that solve problems. Because you need a module to provide values? What happens if the module call `turbine.getExtensionSettings()`, which need to resolve tokens? It's a vicious circle (like we call it in french: ["C'est le serpent qui se mord la queue"](https://en.wiktionary.org/wiki/c%27est_le_serpent_qui_se_mord_la_queue))

### Browsable data element

Add an option (in the manifest or in [data element's attributes](https://developer.adobelaunch.com/api/reference/1.0/data_elements/create/)) that allow data element to be "browsable" with tokens, like [custom vars, `event`, `target` and `this`](https://github.com/adobe/reactor-turbine/blob/2eec8bdb3ed6bd43f8bab53cfeaddfb5e5a6f34b/src/createGetVar.js#L86-L105). Why declare hundred of data elements in your property, when you need only one?

### Allow Launch users to define extensions hydratation order

Actually I'm not sure how extensions are ordered in the libray package. It looks like it's based on the revision date and/or in the library's resources list order.

### Centralized rules/data elements

Some way to reduce the number of data elements to maintain cross multiple properties, especially when there is lot of Launch resources is to have a library property that can be used as a "base" by an other property (kind of "extends").

See the AEPL ideas:

- [Create common centralised rules/data elements shared across properties](https://experienceleaguecommunities.adobe.com/t5/adobe-experience-platform-launch/create-common-centralised-rules-data-elements-shared-across/idi-p/373891)
- [Add master-child management capabilities to Launch](https://experienceleaguecommunities.adobe.com/t5/adobe-experience-platform-launch/add-master-child-management-capabilities-to-launch/idi-p/373850)

In the past this idea has already been mention in a comment, about the idea add the feature to copy resources into other properties:

> the need for a central place to modify rule components and data elements that can be used by multiple properties and managed in one place
> 
> — [Copy/Replicate Adobe Launch Properties (Templates)](https://experienceleaguecommunities.adobe.com/t5/adobe-experience-platform-launch/copy-replicate-adobe-launch-properties-templates/idc-p/341165/highlight/true#M422)

**Comment by mems** [comment]
The only way I found to workaround that issue is to inject in the page that script.

It catch the libray declaration and inject a module that will be the first to be hydrated:

```js
window._satellite = {
  set container(container){
    // can inject new data elements to container.dataElements

    // inject a fake extension that will be hydrated first, iteration respect insertion order since ES2015
    container.extensions = Object.assign({
      "my-fake-extension": {
        displayName: "My fake extension",
        modules: {
          "my-fake-extension/main": {
            // https://github.com/adobe/reactor-turbine/blob/2eec8bdb3ed6bd43f8bab53cfeaddfb5e5a6f34b/src/extractModuleExports.js#L18
            script: function(module, exports, require, turbine) {
              _satellite.setVar({
                window: window,
                digitalData: window.digitalData
              });
            }
          }
        },
        //hostedLibFilesBaseUrl: "https://somewhere"
      }
    }, container.extensions);
    
    Object.defineProperty(this, "container", {value: container, configurable: true, enumerable: true, writable: true});// override the inital setter
  }
}
```

**Comment by thebenrob** [comment]
We discussed briefly, the team finds this to be an interesting idea.  We're going to think about this one a bit and see if we can decide what we want to do with it.

**Comment by thebenrob** [comment]
@adobe export issue to Jira project PDCL as Task

**Comment by brenthosie** [comment]
If this is still relevant, we can discuss again. But this is extremely stale and not on anybody's roadmap.

**Comment by mems** [comment]
> If this is still relevant, we can discuss again. But this is extremely stale and not on anybody's roadmap.

Yes, it's still the case. We have to use a workaround that force extensions order, that I'm not a big fan it. Or we have to create hundreds of data elements.

I think the easier way would be to implement an option for data element to be "browsable" (see [my first comment](https://github.com/adobe/reactor-turbine/issues/116#issue-747637060)), like `event`, `target` or `this` are.
With a browsable data element `window` = `{"a": {"b": "c"}}`, `window.a.b` will give `"c"`
And with support of "escaped" sequence in "query", for DE `window` = `{"a": {"b.c": {"d": "e"}}}`, `window.a.b\.c.d` will give `"e"`

**Comment by mems** [comment]
An other possibility is to add a flag to enable "browsable" syntax, like #205 introduce only where it's needed.
